src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi

```diff
@@ -43,12 +44,49 @@ cdef class RPCState:
             grpc_call_unref(self.call)
 
 
+cdef class _ServicerContext:
+    cdef RPCState _rpc_state
+    cdef object _loop
+    cdef bint _metadata_sent
+    cdef object _request_deserializer
+    cdef object _response_serializer
+
+    def __cinit__(self,
+                  RPCState rpc_state,
+                  object request_deserializer,
+                  object response_serializer,
+                  object loop):
+        self._rpc_state = rpc_state
+        self._request_deserializer = request_deserializer
+        self._response_serializer = response_serializer
+        self._loop = loop
+        self._metadata_sent = False
+
+    async def read(self):
+        cdef bytes raw_message = await _receive_message(self._rpc_state, self._loop)
+        return deserialize(self._request_deserializer,
+                           raw_message)
+
+    async def write(self, object message):
+        await _send_message(self._rpc_state,
+                            serialize(self._response_serializer, message),
+                            self._metadata_sent,
+                            self._loop)
+        if not self._metadata_sent:
+            self._metadata_sent = True
+
+    async def send_initial_metadata(self, tuple metadata):
+        if self._metadata_sent:
+            raise ValueError('Send initial metadata failed: already sent')
```

Comment:
I don't think `ValueError` is quite right here. It's not the value of an argument that's wrong. It's the usage of this function. If `UsageError` were a built-in exception, I would recommend that. Perhaps `RuntimeError` or a subclass of it?